pkg/ibmcsidriver/controller.go

```diff
@@ -418,30 +434,161 @@ func (csiCS *CSIControllerServer) ControllerGetCapabilities(ctx context.Context,
-	return nil, commonError.GetCSIError(ctxLogger, commonError.MethodUnimplemented, requestID, nil, "CreateSnapshot")
+	snapshot, err := session.GetSnapshotByName(snapshotName)
+	if snapshot != nil {
+		if snapshot.VolumeID != sourceVolumeID {
+			return nil, commonError.GetCSIError(ctxLogger, commonError.SnapshotAlreadyExists, requestID, err, snapshotName, sourceVolumeID)
+		}
+		ctxLogger.Info("Snapshot with name already exist for volume", zap.Reflect("SnapshotName", snapshotName), zap.Reflect("VolumeID", sourceVolumeID))
+		return createCSISnapshotResponse(*snapshot), nil
+	}
+	snapshotParameters := provider.SnapshotParameters{}
+	snapshotParameters.Name = snapshotName
+	snapshotTags := map[string]string{
+		"name": snapshotName,
+	}
+	snapshotParameters.SnapshotTags = snapshotTags
+
+	snapshot, err = session.CreateSnapshot(sourceVolumeID, snapshotParameters)
+
+	if err != nil {
+		return nil, commonError.GetCSIError(ctxLogger, commonError.InternalError, requestID, err, "creation")
+	}
+	return createCSISnapshotResponse(*snapshot), nil
 }
 
 // DeleteSnapshot ...
 func (csiCS *CSIControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
 	ctxLogger, requestID := utils.GetContextLogger(ctx, false)
 	// populate requestID in the context
-	_ = context.WithValue(ctx, provider.RequestID, requestID)
+	ctx = context.WithValue(ctx, provider.RequestID, requestID)
+	defer metrics.UpdateDurationFromStart(ctxLogger, "DeleteSnapshot", time.Now())
+	ctxLogger.Info("CSIControllerServer-DeleteSnapshot... ", zap.Reflect("Request", *req))
+
+	// Validate arguments
+	snapshotID := req.GetSnapshotId()
+	if len(snapshotID) == 0 {
+		return nil, commonError.GetCSIError(ctxLogger, commonError.EmptySnapshotID, requestID, nil)
+	}
+
+	// get the session
+	session, err := csiCS.CSIProvider.GetProviderSession(ctx, ctxLogger)
+	if err != nil {
+		if userError.GetUserErrorCode(err) == string(utilReasonCode.EndpointNotReachable) {
+			return nil, commonError.GetCSIError(ctxLogger, commonError.EndpointNotReachable, requestID, err)
+		}
+		if userError.GetUserErrorCode(err) == string(utilReasonCode.Timeout) {
+			return nil, commonError.GetCSIError(ctxLogger, commonError.Timeout, requestID, err)
+		}
+		return nil, commonError.GetCSIError(ctxLogger, commonError.InternalError, requestID, err)
+	}
 
-	ctxLogger.Info("CSIControllerServer-DeleteSnapshot", zap.Reflect("Request", *req))
-	return nil, commonError.GetCSIError(ctxLogger, commonError.MethodUnimplemented, requestID, nil, "DeleteSnapshot")
+	_, err = session.GetSnapshot(snapshotID)
+	if providerError.RetrivalFailed == providerError.GetErrorType(err) {
+		ctxLogger.Info("Snapshot not found. Returning success without deletion...")
+		return &csi.DeleteSnapshotResponse{}, nil
+	}
+	snapshot := &provider.Snapshot{}
+	snapshot.SnapshotID = snapshotID
+	err = session.DeleteSnapshot(snapshot)
+	if err != nil {
```

Comment:
check the error here if snapshot not found and then return

Reply:
Code Coverage

Reply:
sanity improvement

Reply:
The GetSnapshot() method called in DeleteSnapshot() is valid and the explanation to this is :- 
If we call DeleteSnapshot( ) with not present snapshotID then we get a errorCode `FailedToDeleteSnapshot` with the message  like `"BackendError":"Trace Code:eb790aac-899a-4916-adfd-e25ea97c7284, Snapshot with ID 'r134-125d254f-b5b4-4d97-acef-5717f6491b53' is not found Please check Try to list the snapshots and see if the snapshot ID you are trying to find is in the list.","RC":500,"Action":"Check whether the snapshot ID exists. You may need to verify by using 'ibmcloud is' cli"}}`
But the error code `FailedToDeleteSnapshot` is valid because we are calling a delete operation.
Now if we call DeleteSnapshot directly then we need to parse the error string which is not looks to be correct to check if it snapshot not found or some other issue while deleting and then accordingly set the response.
As while deleting CSI expects
- If snapshot is not found -> empty deleteSnapshotResponse and nil error.
- If snapshot is present and some issue while deletion then nil response and the proper error `InternalError`.
And also DeleteVolume has similar implementation.

Reply:
Done